Topic 1
=====

Introduction, no hands on exercises this meeting.

Topic 2
=====

[PRINSEQ](http://prinseq.sourceforge.net) was used for quality trimming and data analysis (link to standalone version is [here](http://sourceforge.net/projects/prinseq/files/standalone)).

Topic 3
=====

Alignment programs used are [bowtie2](http://sourceforge.net/projects/bowtie-bio/files/bowtie2/) and [bwa-mem](https://github.com/lh3/bwa). Also required are [samtools](http://www.htslib.org/) and the R package [rbamtools](https://r-forge.r-project.org/projects/rbamtools/).

Topic 4
=====

[Velvet](https://github.com/dzerbino/velvet) is required for assembly.

Topic 5
=====

[RSEM](http://deweylab.biostat.wisc.edu/rsem/) and bowtie2 (see above) are required, in addition to an R script provided on the course site.

Topic 6
=====

SAMtools (see above), [bcftools](https://github.com/samtools/bcftools) and the R package [VariantAnnotation](http://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html) are used. Note that this R package can be quite difficult to install.

Topic 7
=====

No additional programs are required for this exercise.

Topic 8
=====

[BLAST2GO](https://blast2go.com/blast2go-pro/download-b2g) was used for GO annotation, along with [HMMER2GO](https://github.com/sestaton/HMMER2GO). [Ontologizer](http://compbio.charite.de/contao/index.php/ontologizer2.html) is used for GO term enrichment. Also, a Bash script was provided to automate the GO annotation process.

Topic 9
=====

All exercises are done with the [Seaview](http://doua.prabi.fr/software/seaview) application.

Topic 10
=====

Wrap up discussion and lecture on phylogenomics, no hands on exercises.